proposals/0000-out-of-npm-artifacts.md

```diff
@@ -0,0 +1,337 @@
+
+Github offers the possibility to use the [Container Registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry) to upload archives without size limits (for public repositories), as long as they conform to the [OCI (Open Container Initiative) specs](https://opencontainers.org/posts/announcements/2021-05-04-oci-dist-spec-v1/#:~:text=The%20OCI%20Distribution%20Spec%20defines,and%20pulling%20of%20container%20images.).  This is the same mechanism used by Brew to distribute binaries ([source](https://github.blog/2021-06-21-github-packages-container-registry-generally-available/#dont-just-take-our-word-for-it)).
+
+Ideally we could create a container, which separate layer per artifact (i.e. one for Hermes iOS tarball, one for React Native Android, etc). Each layer will be a .tar.gz file we can freely populate.
+
+#### Pro
+
+* Elegant solution which relies on the Github infrastructure and doens’t involve external services (no need to auth, sign, or interact with new services).
+* No limits to how much we can upload.
+* Potentially we could access the Container Registry API to allow users to download only the layer they need.
+
+#### Cons
+
+* Need to write extra infrastructure to handle downloading the OCI images and placing the artifacts in the right folder.
+* No caching mechanism out of the box (users with 2 React Native projects, will re-download everything). 
+
+### Separate NPM Package
+
+Ideally we could look into splitting `react-native` into smaller packages:
+
+* `@react-native/android-prebuilts` will contain the `android/` folder
+* `@react-native/hermesc-prebuilts` will contain the `sdks/hermes` folder
+
+This will partially mitigate the issue as the `android-prebuilts` package will still hit the NPM limit once we include debug symbols. This will require further splitting such as:
+
+* Per build-type split `android-debug-prebuilts` and `android-release-prebuilts`
+* Per library split `react-native-prebuilts` and `hermes-engine-prebuilts`
+* Per ABI split `x86_64-prebuilts`, `arm64-v8a-prebuilts` and so on.
+
+#### Pro
+
+* Uses a packaging system which is familiar for web developers
+* We should be able to reuse most of the publishing pipeline we have in place at the moment
+
+#### Cons
+
+* We will be forced to “recompose” the various prebuilts from the several NPM packages to a Maven Local repository inside users’ `node_modules` folder.
+* All the users will be downloading all the pre-builts for all the build-types/split/abi whenever they `yarn install`.
+* Having [github.com/facebook/react-native](http://github.com/facebook/react-native) as currently set up (not a proper monorepo) makes really hard to create new NPM packages.
+* We will still be having a hard limit on NPM packages so we’ll have to handle further splitting as our library/SDK grows. This is especially more critical as we have more and more C++ code, therefore we’re generating more and more native pre-builts.
```

Comment:
Another downside is that users who develop for a single platform incur a download cost for binaries they will never use, vs delaying the download until the specific native build system needs the artifact.

Reply:
Yup I'll expand on this point 👍

Reply:
Done in [f5f2f50](https://github.com/react-native-community/discussions-and-proposals/pull/508/commits/f5f2f501242f4f1b90f4d718a48834f6404a9bdc)